File: src/modules/b2bProposal.js
/**
 * Rayeva AI Systems – Module 2: AI B2B Proposal Generator
 *
 * Business Logic:
 * - Accepts client details (name, industry, budget, requirements)
 * - Calls AI to generate sustainable product proposal
 * - Validates budget constraints
 * - Stores structured result in database
 * - Logs the AI interaction
 */

const aiClient = require("../ai/client");
const {
  B2B_PROPOSAL_SYSTEM,
  buildProposalUserPrompt,
} = require("../ai/prompts");
const { logAiInteraction } = require("../ai/aiLogger");
const config = require("../config");
const { B2BProposal } = require("../models");
const logger = require("../ai/logger");

const MODULE_NAME = "b2b_proposal";

/**
 * Generate a B2B proposal using AI and store result.
 * @param {object} params
 * @param {string} params.clientName
 * @param {string} params.clientIndustry
 * @param {number} params.budgetLimit
 * @param {string|null} [params.requirements]
 * @param {string|null} [params.sustainabilityPriority]
 * @returns {Promise<object>} Proposal data with AI-generated content
 */
async function generateProposal({
  clientName,
  clientIndustry,
  budgetLimit,
  requirements = null,
  sustainabilityPriority = "medium",
}) {
  // 1. Build prompt
  const userPrompt = buildProposalUserPrompt(
    clientName,
    clientIndustry,
    budgetLimit,
    requirements,
    sustainabilityPriority
  );

  // 2. Call AI with higher token limit for detailed proposals
  const aiResult = await aiClient.generate(
    B2B_PROPOSAL_SYSTEM,
    userPrompt,
    0.4, // Slightly creative for proposals
    3000
  );

  // 3. Log the interaction
  await logAiInteraction({
    module: MODULE_NAME,
    prompt: `SYSTEM: ${B2B_PROPOSAL_SYSTEM}\n\nUSER: ${userPrompt}`,
    response: aiResult.response_text,
    model_used: config.OPENAI_MODEL,
    tokens_used: aiResult.tokens_used,
    latency_ms: aiResult.latency_ms,
    success: aiResult.success,
    error_message: aiResult.error,
  });

  // 4. Handle AI failure
  if (!aiResult.success || !aiResult.parsed_json) {
    throw new Error(
      `AI proposal generation failed: ${aiResult.error || "Unknown error"}`
    );
  }

  // 5. Validate and clean the AI output
  const rawJson = aiResult.parsed_json;
  const validated = validateOutput(rawJson, budgetLimit);

  // 6. Store proposal in database
  const proposal = await B2BProposal.create({
    client_name: clientName,
    client_industry: clientIndustry,
    budget_limit: budgetLimit,
    requirements,
    product_mix: validated.product_mix,
    budget_allocation: validated.budget_allocation,
    cost_breakdown: validated.cost_breakdown,
    impact_summary: validated.impact_summary,
    full_ai_response: rawJson,
    ai_generated_at: new Date(),
  });

  logger.info(
    `B2B Proposal generated: id=${proposal.id}, client=${clientName}, ` +
      `total=₹${validated.cost_breakdown.total_estimated.toLocaleString("en-IN", { minimumFractionDigits: 2 })}`
  );

  // 7. Return structured response
  return {
    id: proposal.id,
    client_name: proposal.client_name,
    client_industry: proposal.client_industry,
    budget_limit: proposal.budget_limit,
    requirements: proposal.requirements,
    proposal: validated,
    created_at: proposal.created_at,
    ai_generated_at: proposal.ai_generated_at,
  };
}

/**
 * Validate AI proposal output against business rules.
 * - Ensures total cost doesn't exceed budget
 * - Validates product mix structure
 * - Recalculates totals if needed
 * @param {object} raw
 * @param {number} budgetLimit
 * @returns {object} Validated proposal
 */
function validateOutput(raw, budgetLimit) {
  // Parse product mix
  const productMix = [];
  for (const item of raw.product_mix || []) {
    try {
      const unitPrice = parseFloat(item.unit_price_estimate) || 0;
      const quantity = parseInt(item.quantity, 10) || 0;
      const calculatedTotal = Math.round(unitPrice * quantity * 100) / 100;

      productMix.push({
        product_name: item.product_name || "Unknown Product",
        category: item.category || "General",
        unit_price_estimate: unitPrice,
        quantity,
        total_price: calculatedTotal,
        sustainability_tags: item.sustainability_tags || [],
      });
    } catch (err) {
      logger.warn(`Skipping invalid product mix item: ${err.message}`);
    }
  }

  // Recalculate actual product costs
  let actualProductCosts = productMix.reduce(
    (sum, item) => sum + item.total_price,
    0
  );

  // Parse budget allocation
  const budgetAllocation = [];
  for (const alloc of raw.budget_allocation || []) {
    try {
      budgetAllocation.push({
        category: alloc.category || "General",
        allocated_amount: parseFloat(alloc.allocated_amount) || 0,
        percentage: parseFloat(alloc.percentage) || 0,
      });
    } catch (err) {
      logger.warn(`Skipping invalid budget allocation: ${err.message}`);
    }
  }

  // Parse and validate cost breakdown
  const rawBreakdown = raw.cost_breakdown || {};
  let shipping =
    parseFloat(rawBreakdown.shipping_estimate) || actualProductCosts * 0.08;
  let packaging =
    parseFloat(rawBreakdown.packaging_costs) || actualProductCosts * 0.04;
  let totalEstimated =
    Math.round((actualProductCosts + shipping + packaging) * 100) / 100;

  // Budget enforcement: scale down if over budget
  if (totalEstimated > budgetLimit) {
    const scaleFactor = (budgetLimit / totalEstimated) * 0.95; // 5% buffer
    logger.warn(
      `Proposal exceeds budget (₹${totalEstimated.toLocaleString()} > ₹${budgetLimit.toLocaleString()}). ` +
        `Scaling down by factor ${scaleFactor.toFixed(2)}`
    );

    for (const item of productMix) {
      item.quantity = Math.max(1, Math.floor(item.quantity * scaleFactor));
      item.total_price =
        Math.round(item.unit_price_estimate * item.quantity * 100) / 100;
    }

    actualProductCosts = productMix.reduce(
      (sum, item) => sum + item.total_price,
      0
    );
    shipping = Math.round(actualProductCosts * 0.08 * 100) / 100;
    packaging = Math.round(actualProductCosts * 0.04 * 100) / 100;
    totalEstimated =
      Math.round((actualProductCosts + shipping + packaging) * 100) / 100;
  }

  const remaining = Math.round((budgetLimit - totalEstimated) * 100) / 100;

  const costBreakdown = {
    product_costs: Math.round(actualProductCosts * 100) / 100,
    shipping_estimate: Math.round(shipping * 100) / 100,
    packaging_costs: Math.round(packaging * 100) / 100,
    total_estimated: totalEstimated,
    remaining_budget: Math.max(0, remaining),
  };

  // Impact summary
  const impactSummary =
    raw.impact_summary ||
    "Sustainable procurement proposal for your organization.";
  const keySellingPoints = raw.key_selling_points || [
    "Reduced environmental footprint",
    "Enhanced brand sustainability positioning",
    "Cost-effective green alternatives",
  ];

  return {
    product_mix: productMix,
    budget_allocation: budgetAllocation,
    cost_breakdown: costBreakdown,
    impact_summary: impactSummary,
    key_selling_points: keySellingPoints,
  };
}

module.exports = { generateProposal };

File: rayeva-ai-systems/app/modules/b2b_proposal.py
"""
Rayeva AI Systems – Module 2: AI B2B Proposal Generator

Business Logic:
- Accepts client details (name, industry, budget, requirements)
- Calls AI to generate sustainable product proposal
- Validates budget constraints
- Stores structured result in database
- Logs the AI interaction
"""

import logging
from datetime import datetime, timezone
from sqlalchemy.orm import Session

from app.ai.client import ai_client
from app.ai.prompts import B2B_PROPOSAL_SYSTEM, build_proposal_user_prompt
from app.ai.logger import log_ai_interaction
from app.config import get_settings
from app.models import B2BProposal
from app.schemas import ProposalResult, ProductMixItem, BudgetAllocation, CostBreakdown

logger = logging.getLogger(__name__)
settings = get_settings()


class B2BProposalGenerator:
    """
    AI-powered B2B proposal generation engine.

    Responsibilities:
    - Call AI with client details and budget
    - Validate that proposal stays within budget
    - Persist to database
    - Return structured proposal
    """

    MODULE_NAME = "b2b_proposal"

    def generate_proposal(
        self,
        db: Session,
        client_name: str,
        client_industry: str,
        budget_limit: float,
        requirements: str | None = None,
        sustainability_priority: str | None = "medium",
    ) -> dict:
        """
        Generate a B2B proposal using AI and store result.

        Args:
            db: Database session
            client_name: Client/company name
            client_industry: Industry vertical
            budget_limit: Budget limit in INR
            requirements: Specific client requirements
            sustainability_priority: low / medium / high

        Returns:
            dict with proposal data and AI-generated content
        """
        # 1. Build prompt
        user_prompt = build_proposal_user_prompt(
            client_name, client_industry, budget_limit,
            requirements, sustainability_priority
        )

        # 2. Call AI with higher token limit for detailed proposals
        ai_result = ai_client.generate(
            system_prompt=B2B_PROPOSAL_SYSTEM,
            user_prompt=user_prompt,
            temperature=0.4,  # Slightly creative for proposals
            max_tokens=3000,
        )

        # 3. Log the interaction
        log_ai_interaction(
            db=db,
            module=self.MODULE_NAME,
            prompt=f"SYSTEM: {B2B_PROPOSAL_SYSTEM}\n\nUSER: {user_prompt}",
            response=ai_result["response_text"],
            model_used=settings.OPENAI_MODEL,
            tokens_used=ai_result["tokens_used"],
            latency_ms=ai_result["latency_ms"],
            success=ai_result["success"],
            error_message=ai_result["error"],
        )

        # 4. Handle AI failure
        if not ai_result["success"] or not ai_result["parsed_json"]:
            raise ValueError(
                f"AI proposal generation failed: {ai_result['error'] or 'Unknown error'}"
            )

        # 5. Validate and clean the AI output
        raw_json = ai_result["parsed_json"]
        validated = self._validate_output(raw_json, budget_limit)

        # 6. Store proposal in database
        proposal = B2BProposal(
            client_name=client_name,
            client_industry=client_industry,
            budget_limit=budget_limit,
            requirements=requirements,
            product_mix=[item.model_dump() for item in validated.product_mix],
            budget_allocation=[alloc.model_dump() for alloc in validated.budget_allocation],
            cost_breakdown=validated.cost_breakdown.model_dump(),
            impact_summary=validated.impact_summary,
            full_ai_response=raw_json,
            ai_generated_at=datetime.now(timezone.utc),
        )
        db.add(proposal)
        db.commit()
        db.refresh(proposal)

        logger.info(
            f"B2B Proposal generated: id={proposal.id}, client={client_name}, "
            f"total=₹{validated.cost_breakdown.total_estimated:,.2f}"
        )

        # 7. Return structured response
        return {
            "id": proposal.id,
            "client_name": proposal.client_name,
            "client_industry": proposal.client_industry,
            "budget_limit": proposal.budget_limit,
            "requirements": proposal.requirements,
            "proposal": validated.model_dump(),
            "created_at": proposal.created_at.isoformat(),
            "ai_generated_at": proposal.ai_generated_at.isoformat(),
        }

    def _validate_output(self, raw: dict, budget_limit: float) -> ProposalResult:
        """
        Validate AI proposal output against business rules.

        - Ensures total cost doesn't exceed budget
        - Validates product mix structure
        - Recalculates totals if needed
        """
        # Parse product mix
        product_mix = []
        for item in raw.get("product_mix", []):
            try:
                # Recalculate total_price to ensure consistency
                unit_price = float(item.get("unit_price_estimate", 0))
                quantity = int(item.get("quantity", 0))
                calculated_total = round(unit_price * quantity, 2)

                product_mix.append(ProductMixItem(
                    product_name=item.get("product_name", "Unknown Product"),
                    category=item.get("category", "General"),
                    unit_price_estimate=unit_price,
                    quantity=quantity,
                    total_price=calculated_total,
                    sustainability_tags=item.get("sustainability_tags", []),
                ))
            except (ValueError, TypeError) as e:
                logger.warning(f"Skipping invalid product mix item: {e}")
                continue

        # Recalculate actual product costs
        actual_product_costs = sum(item.total_price for item in product_mix)

        # Parse budget allocation
        budget_allocation = []
        for alloc in raw.get("budget_allocation", []):
            try:
                budget_allocation.append(BudgetAllocation(
                    category=alloc.get("category", "General"),
                    allocated_amount=float(alloc.get("allocated_amount", 0)),
                    percentage=float(alloc.get("percentage", 0)),
                ))
            except (ValueError, TypeError) as e:
                logger.warning(f"Skipping invalid budget allocation: {e}")
                continue

        # Parse and validate cost breakdown
        raw_breakdown = raw.get("cost_breakdown", {})
        shipping = float(raw_breakdown.get("shipping_estimate", actual_product_costs * 0.08))
        packaging = float(raw_breakdown.get("packaging_costs", actual_product_costs * 0.04))
        total_estimated = round(actual_product_costs + shipping + packaging, 2)

        # Budget enforcement: scale down if over budget
        if total_estimated > budget_limit:
            scale_factor = budget_limit / total_estimated * 0.95  # 5% buffer
            logger.warning(
                f"Proposal exceeds budget (₹{total_estimated:,.2f} > ₹{budget_limit:,.2f}). "
                f"Scaling down by factor {scale_factor:.2f}"
            )
            for item in product_mix:
                item.quantity = max(1, int(item.quantity * scale_factor))
                item.total_price = round(item.unit_price_estimate * item.quantity, 2)

            actual_product_costs = sum(item.total_price for item in product_mix)
            shipping = round(actual_product_costs * 0.08, 2)
            packaging = round(actual_product_costs * 0.04, 2)
            total_estimated = round(actual_product_costs + shipping + packaging, 2)

        remaining = round(budget_limit - total_estimated, 2)

        cost_breakdown = CostBreakdown(
            product_costs=round(actual_product_costs, 2),
            shipping_estimate=round(shipping, 2),
            packaging_costs=round(packaging, 2),
            total_estimated=total_estimated,
            remaining_budget=max(0, remaining),
        )

        # Impact summary
        impact_summary = raw.get("impact_summary", "Sustainable procurement proposal for your organization.")
        key_selling_points = raw.get("key_selling_points", [
            "Reduced environmental footprint",
            "Enhanced brand sustainability positioning",
            "Cost-effective green alternatives",
        ])

        return ProposalResult(
            product_mix=product_mix,
            budget_allocation=budget_allocation,
            cost_breakdown=cost_breakdown,
            impact_summary=impact_summary,
            key_selling_points=key_selling_points,
        )


# Singleton
b2b_proposal_generator = B2BProposalGenerator()
